grep/cbeer/grep.go

```diff
@@ -0,0 +1,118 @@
+package grep
+
+import (
+	"bufio"
+	"fmt"
+	"log"
+	"os"
+	"strings"
+)
+
+// Search a file for lines matching a regular expression pattern.
+func Search(pattern string, flags, files []string) []string {
+	searchResults := make([]string, 0)
+	var printLineNumbers bool
+	var printOnlyNames bool
+	var includeFileNames bool
+
+	for _, flagValue := range flags {
+		switch flagValue {
+		case "-n":
+			printLineNumbers = true
+		case "-l":
+			printOnlyNames = true
+		}
+	}
+
+	if len(files) > 1 {
+		includeFileNames = true
+	}
+
+	for _, fileName := range files {
+		fileHandle, err := os.Open(fileName)
+		if err != nil {
+			log.Fatal(err)
+		}
+		defer fileHandle.Close()
+
+		scanner := bufio.NewScanner(fileHandle)
+		lineNumber := 1
+		var lineOfText string
+
+		if printOnlyNames {
+			matchFound := false
+			for scanner.Scan() {
+				if isLineMatch(pattern, scanner.Text(), flags) {
+					matchFound = true
+				}
+			}
+			if matchFound {
+				searchResults = append(searchResults, fileName)
+			}
+		} else {
+			for scanner.Scan() {
+				if isLineMatch(pattern, scanner.Text(), flags) {
+					if includeFileNames {
+						if printLineNumbers {
+							lineOfText = fmt.Sprintf("%s:%d:%s", fileName, lineNumber, scanner.Text())
```

Comment:
You can have scanner.Text() called just once to return a line, then pass it to the matcher function and modify it for output instead of using scanner.Text() which is less readable than a var `line`.
And you can have less cases if you check once for `-n` and prepend it to line and again for number of files and again prepend the filename to the line.